snuba/query/processors/tags_promoter.py

```diff
@@ -0,0 +1,105 @@
+from snuba.clickhouse.query import Query
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
+from snuba.request.request_settings import RequestSettings
+
+
+class PromotedColumnsSpec(NamedTuple):
+    # The name of the key field in the mapping column
+    key_field: str
+    # The name of the value field in the mapping column
+    val_field: str
+    # The mapping between keys in the column and real columns
+    column_mapping: Mapping[str, str]
+
+
+MappingColumnPromotionSpec = Mapping[str, PromotedColumnsSpec]
+
+
+class MappingColumnPromoter(QueryProcessor):
+    """
+    Promotes expression to access the value of a mapping column by replacing them with
+    the corresponding promoted column.
+
+    Example: tags["myTag"] -> arrayElement("tags.value", indexOf("tags.key", "myTag"))
+     -> toString(promoted_MyTag)
+
+    This happens if there is a promoted_MyTag column in the storage that maps to the myTag tag.
+    """
+
+    def __init__(
+        self, columns: ColumnSet, mapping_spec: MappingColumnPromotionSpec
+    ) -> None:
+        # The ColumnSet of the dataset. Used to format promoted
+        # columns with the right type.
+        self.__columns = columns
+        # Maps nested column keys to real column names.
+        self.__spec = mapping_spec
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        def transform_nested_column(exp: Expression) -> Expression:
+            # TODO: use the matcher system developed in #942 when merged to make this
+            # less verbose and put the code that generates this expression (in #936) in
+            # the same place as the matcher expression so they do not diverge.
+            #
+            # Now it identifies arrayElement("tags.value", indexOf("tags.key", "myTag"))
+            if not (
+                isinstance(exp, FunctionCall)
+                and exp.function_name == "arrayElement"
+                and len(exp.parameters) == 2
+                and isinstance(exp.parameters[0], Column)
+                and isinstance(exp.parameters[1], FunctionCall)
+                and exp.parameters[1].function_name == "indexOf"
+                and len(exp.parameters[1].parameters) == 2
+                and isinstance(exp.parameters[1].parameters[0], Column)
+                and isinstance(exp.parameters[1].parameters[1], Literal)
+            ):
+                return exp
+
+            # TODO: There is should be a structured Column class to deal with references
+            # to nested columns instead of splitting the column name string. Resolution
+            # of such column is tightly coupled to how we will do entity resolution.
+            val_column = exp.parameters[0]
+            val_column_splits = val_column.column_name.split(".", 2)
+
+            key_column = exp.parameters[1].parameters[0]
+            key_column_splits = key_column.column_name.split(".", 2)
```

Comment:
I want to make an attempt at resolving nested column expressions early in the query process before relying on string split to achieve this.

Reply:
Here is is
https://github.com/getsentry/snuba/pull/963